src/lib/Libifl/pbsD_connect.c

```diff
@@ -364,45 +416,18 @@ get_conn_svr_instances(int parentfd)
 {
 	svr_conns_list_t *conn_list = NULL;
 	svr_conn_t *msvr_conns = NULL;
+	svr_conns_list_t *new_conns = NULL;
 
 	conn_list = pthread_getspecific(psi_key);
-	if (conn_list == NULL || parentfd == -1) {	/* Create a new connection set */
-		int num_svrs;
-		int i;
-		svr_conns_list_t *new_conns = NULL;
-
-		new_conns = malloc(sizeof(svr_conns_list_t));
-		if (new_conns == NULL) {
-			pbs_errno = PBSE_SYSTEM;
-			return NULL;
-		}
-		new_conns->next = NULL;
-
-		num_svrs = get_num_servers();
-		msvr_conns = calloc(num_svrs, sizeof(svr_conn_t));
-		if (msvr_conns == NULL) {
-			free(new_conns);
-			pbs_errno = PBSE_SYSTEM;
-			return NULL;
-		}
-
-		for (i = 0; i < num_svrs; i++) {
-			strcpy(msvr_conns[i].name, pbs_conf.psi[i].name);
-			msvr_conns[i].port = pbs_conf.psi[i].port;
-			msvr_conns[i].sd = -1;
-			msvr_conns[i].state = SVR_CONN_STATE_DOWN;
-		}
-		new_conns->conn_arr = msvr_conns;
-		new_conns->next = conn_list;
-		conn_list = new_conns;
-
-		pthread_setspecific(psi_key, conn_list);
+	if (conn_list == NULL) { /* Create a new connection set */
```

Comment:
There's a reason why the parentfd == -1 check existed. From what I understand, this change can lead to the following:
- client calls pbs_connect->connect_to_servers() the first time, conn_list is NULL so this function will create a new list and a new svr_conn_t array, allocate it and send it back
- client calls pbs_connect->connect_to_servers() again (like we do in sched), this time, conn_list is not NULL, so this if condition will end up not creating a new connection list, or a new svr_conn_t array, and return NULL? Or have I misunderstood this?

Reply:
I haven't changed the creation logic. I've refactored the part where the conn array is created into a different function. So the place get_conn_svr_instances was invoked with -1 is replaced by a call to create_conn_svr_instances.
If the creation happens only via pbs_connect, we shouldn't be creating inside get_conn_svr_instances at all. Client relies on this routine to iterate through connections, so calling this for error logging etc will be just creating extra connection structures.

Reply:
I understand, but i think your refactoring has changed the logic. When you call create_conn_svr_instances() the second time, it will not create a new svr_conn_t because conn_list will not be NULL.

Reply:
create_conn_svr_instances will create a new connection irrespective of conditions. The new commit contains more pbs_connect changes.

Reply:
Looks good now, and the refactoring does make a lot of sense, thanks.